### PR TITLE
os: implement readdir for darwin and linux

### PR DESCRIPTION
**Pull request by dkegel-fastly**
Other operating systems are left as an exercise for the reader :-)

### CODE REVIEW

**Comment by dkegel-fastly** [comment]
The i386 problem this ran into in CI may be related to https://github.com/tinygo-org/tinygo/issues/1906

It can be reproduced with the commands

```
make
cd tests/testing/pass
GOARCH=386 ../../../build/tinygo test
```
in this branch, but not in dev.

**Review by dgryski — APPROVED** [review]
LGTM.  Large patch, but almost all of it seems to be from upstream Go. And it has tests, which pass.  👍

**Comment by deadprogram** [comment]
Something not working quite right here: https://github.com/tinygo-org/tinygo/runs/4845686957?check_suite_focus=true#step:17:420

```
tinygo:ld.lld: error: undefined symbol: syscall.seek
```

**Comment by dkegel-fastly** [comment]
Yes, see https://github.com/tinygo-org/tinygo/pull/2499#issuecomment-1013768071

The real fix may involves Ayke's fix for https://github.com/tinygo-org/tinygo/issues/1906

Until then, maybe I could figure out some clever way to exclude the architectures with the problem...?

**Comment by dkegel-fastly** [comment]
Needs a tiny bit more cleanup.  Also can't land until the go assembly feature lands.  So marking draft again for now.

**Comment by dkegel-fastly** [comment]
Rebased atop https://github.com/tinygo-org/tinygo/pull/2571 for the seek fix

**Comment by dkegel-fastly** [comment]
https://github.com/tinygo-org/tinygo/pull/2571 should land first, but otherwise this is ready for review.

File: src/os/dir_darwin.go
Comment (dkegel-fastly) [inline src/os/dir_darwin.go]: @aykevl syscall.Fdopendir needs reviewing, I think.

Comment (dkegel-fastly) [reply src/os/dir_darwin.go]: sorted.

**Comment by dankegel** [comment]
I think this is ready, now that I've tested it with io/fs and fixed the problem this found on Darwin.

**Comment by dkegel-fastly** [comment]
Oh, ok, maybe it's too late for 0.22.

**Review by dgryski — APPROVED** [review]
LGTM

**Comment by deadprogram** [comment]
Thanks @dkegel-fastly and @dgryski now merging.

File: src/os/file_other.go
Comment (QuLogic) [inline src/os/file_other.go:34]: This differs from Go in taking an interface for `fd` instead of a `uintptr`, which seems to break if passing, e.g., file descriptor 3 as done in [Go 1.18's fuzzing code](https://github.com/golang/go/blob/d588f487703e773ba4a2f0a04f2d4141610bff6b/src/internal/fuzz/sys_posix.go#L61) (cf. #2515)

Comment (dkegel-fastly) [reply src/os/file_other.go:34]: IIRC that was taken straight from upstream?  Maybe upstream has changed, or maybe I'm misremembering.

Either way, please file a bug and we can sort it.

Comment (QuLogic) [reply src/os/file_other.go:34]: Well, it's been that way for 10 years, it appears. I'll just open a PR as I've already made the changes locally.

**Pull request merged**
